src/Workspaces/Core/MSBuild.BuildHost/Rpc/RpcServer.cs

```diff
@@ -0,0 +1,167 @@
+    {
+        Response response;
+
+        try
+        {
+            Contract.ThrowIfFalse(
+                _rpcTargets.TryGetValue(request.TargetObject, out var rpcTarget),
+                $"Received a request for target object {request.TargetObject} but we don't have a registered object for that.");
+
+            var method = rpcTarget.GetType().GetMethod(request.Method, BindingFlags.Public | BindingFlags.Instance);
+
+            Contract.ThrowIfNull(method, $"The invoked method '{request.Method}' could not be found.");
+
+            var methodParameters = method.GetParameters();
+
+            var lastParameterIsCancellationToken = methodParameters.Length > 0 && methodParameters[^1].ParameterType == typeof(CancellationToken);
+
+            if (lastParameterIsCancellationToken)
+                Contract.ThrowIfFalse(request.Parameters.Length == methodParameters.Length - 1, $"The arguments list should contain every parameter for {request.Method} except the final CancellationToken.");
+            else
+                Contract.ThrowIfFalse(request.Parameters.Length == methodParameters.Length, $"The arguments list should contain every parameter for {request.Method}.");
+
+            var arguments = new object?[methodParameters.Length];
+
+            for (var i = 0; i < methodParameters.Length; i++)
+            {
+                // If the method we're calling accepts a cancellation token, we wouldn't have passed that, so add in a CancellationToken.None here. Although we could
+                // remove the cancellation from the underlying method, this keeps that support around should we need it.
+                if (i == methodParameters.Length - 1 && lastParameterIsCancellationToken)
+                    arguments[i] = CancellationToken.None;
+                else
+                    arguments[i] = request.Parameters[i].ToObject(methodParameters[i].ParameterType);
+            }
+
+            var result = method.Invoke(rpcTarget, arguments);
+
+            if (result is Task task)
+            {
+                await task.ConfigureAwait(false);
+
+                // If it's actually a Task<T> then get the result; we're looking at the declared return type because in some cases a method might
+                // return a Task<T> under the covers as a workaround for the lack of TaskCompletionSource on .NET Framework but we don't want to see
+                // that workaround since the result isn't intended to be seen.
+                if (method.ReturnType.IsConstructedGenericType)
+                {
+                    result = task.GetType().GetProperty("Result")!.GetValue(task);
+                }
+                else
+                {
+                    // It's just a simple Task so no result to return
+                    result = null;
+                }
+            }
+
+            response = new Response { Id = request.Id, Value = result is not null ? JToken.FromObject(result) : null };
```

Comment:
Why do we need to explicitly convert to JToken first, instead of passing it in and letting SerializeObject serialize the original vlaue?

Reply:
Response being typed as JToken here is needed on the client side, since the client knows what the target type is that's being expected. For example, if we're serializing a list of things across, we have to tell Newtonsoft.Json that we want an ImmutableArray() or whatever else we want, or it might pick the wrong type. So on the deserialization side, we need this as JToken so we can be explicit about the type. Since Request is used for sending too, it's going to be come a JToken as well.